File: app/src/main/java/com/foxek/simpletimer/ui/interval/IntervalAdapter.java
package com.foxek.simpletimer.ui.interval;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.foxek.simpletimer.R;
import com.foxek.simpletimer.data.model.Interval;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.foxek.simpletimer.utils.IntervalUtils.formatIntervalData;

public class IntervalAdapter extends ListAdapter<Interval, IntervalAdapter.ViewHolder> {

    private Callback        mCallback;

    public IntervalAdapter() {
        super(DIFF_CALLBACK);
    }

    public interface Callback {
        void onListItemClick(Interval item);
    }

    void setCallback(Callback callback) {
        mCallback = callback;
    }

    @Override
    public void onBindViewHolder(@NonNull IntervalAdapter.ViewHolder holder, int position) {

        holder.mWorkIntervalText.setText(formatIntervalData(getItem(position).getWorkTime()));
        holder.mRestIntervalText.setText(formatIntervalData(getItem(position).getRestTime()));
    }

    @NonNull
    @Override
    public IntervalAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.interval_item, parent, false);
        return new IntervalAdapter.ViewHolder(view);
    }

    private static final DiffUtil.ItemCallback<Interval> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<Interval>() {
                @Override
                public boolean areItemsTheSame(
                        @NonNull Interval oldInterval, @NonNull Interval newInterval) {
                    return ((oldInterval.getWorkTime() == newInterval.getWorkTime()) &&
                            (oldInterval.getRestTime() == newInterval.getRestTime()));
                }
                @Override
                public boolean areContentsTheSame(
                        @NonNull Interval oldInterval, @NonNull Interval newInterval) {
                    return ((oldInterval.getWorkTime() == newInterval.getWorkTime()) &&
                            (oldInterval.getRestTime() == newInterval.getRestTime()));
                }
            };

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.work_interval_text)
        TextView mWorkIntervalText;

        @BindView(R.id.rest_interval_text)
        TextView mRestIntervalText;

        ViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
        }

        @OnClick({R.id.interval_item})
        @Override
        public void onClick(View v) {
            mCallback.onListItemClick(getItem(getAdapterPosition()));
        }

    }
}

File: app/src/main/java/com/foxek/simpletimer/ui/interval/adapter/IntervalAdapter.kt
package com.foxek.simpletimer.ui.interval.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.foxek.simpletimer.R
import com.foxek.simpletimer.data.model.Interval
import com.foxek.simpletimer.utils.Constants.EMPTY
import com.foxek.simpletimer.utils.Constants.WITH_REST_TYPE
import com.foxek.simpletimer.utils.formatIntervalData
import com.foxek.simpletimer.utils.formatIntervalNumber
import kotlinx.android.synthetic.main.multi_interval_item.view.*

class IntervalAdapter : ListAdapter<Interval, IntervalAdapter.ViewHolder> (IntervalDiffCallback()) {

    private var mCallback: Callback? = null

    init {
        setHasStableIds(true)
    }

    interface Callback {
        fun onListItemClick(item: Interval)
    }

    internal fun setCallback(callback: Callback) {
        mCallback = callback
    }

    override fun getItemViewType(position: Int): Int {
        return getItem(position).type
    }

    override fun onViewDetachedFromWindow(holder: ViewHolder) {
        super.onViewDetachedFromWindow(holder)

        // small fix
        holder.itemView.post { notifyItemRangeChanged(holder.layoutPosition + 1,itemCount)}
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        holder.itemView.item_interval_work_time.text = formatIntervalData(getItem(position).work)

        if (getItem(position).type == WITH_REST_TYPE) {
            holder.itemView.item_interval_rest_time.text = formatIntervalData(getItem(position).rest)
        }else{
            holder.itemView.item_interval_rest_hint.text = holder.itemView.resources.getString(R.string.timer_without_rest)
        }

        if (getItem(position).name == EMPTY) {
            holder.itemView.item_interval_title.text = formatIntervalNumber(position+1)
        }else{
            holder.itemView.item_interval_title.text = getItem(position).name
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.multi_interval_item, parent, false)
        return ViewHolder(view)
    }


    inner class ViewHolder constructor(v: View) : RecyclerView.ViewHolder(v) {

        init {
            itemView.item.setOnClickListener { mCallback?.onListItemClick(getItem(adapterPosition)) }
        }
    }
}